### PR TITLE
lesson 1 homework. Sulima Denis

### PR DESCRIPTION
**Pull request by SD3nis**

### CODE REVIEW

File: Lesson1/Denis Sulima/task2/CMakeLists.txt.user [hunk 1]
```diff
@@ -0,0 +1,385 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (smay1613) [inline Lesson1/Denis Sulima/task2/CMakeLists.txt.user]: This type of files must be added to gitignore. Also, could you please merge three tasks into one project? Thanks.

File: Lesson1/Denis Sulima/task1/main.cpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
+#include <iostream>
+#include <vector>
+std::vector<int> compoundVectors(std::vector<int>::iterator vectorFirstBegin, std::vector<int>::iterator vectorFirstEnd,
+                                std::vector<int>::iterator vectorSecondBegin, std::vector<int>::iterator vectorSecondEnd) {
+    std::vector<int>tmp {vectorFirstBegin,vectorFirstEnd};
+    tmp.insert(tmp.end(), vectorSecondBegin, vectorSecondEnd);
+    return tmp;
+}
```
Comment (smay1613) [inline Lesson1/Denis Sulima/task1/main.cpp]: Nice approach.
Try to estimate the algorithm complexity by time and memory in big O notation.

File: Lesson1/Denis Sulima/task2/main.cpp [hunk 3]
```diff
@@ -0,0 +1,28 @@
+#include <iostream>
+#include <list>
+
+
+int main()
+{
+    int X;
```
Comment (smay1613) [inline Lesson1/Denis Sulima/task2/main.cpp]: Please, don't leave variables uninitialized. 
You can easily default initialize it with such synax:
int X {};

File: Lesson1/Denis Sulima/task2/main.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    std::list<int>result;
+    int numberDegree = 1;
+    for (int i = 1; i<=X; i++){
+        numberDegree *= N;
+        int tmpFirstNumber = numberDegree;
+        int tmpSecondNubmer = numberDegree % 10;
+        while (tmpFirstNumber > 9) {
+            tmpFirstNumber /= 10;
+        }
+        result.push_back(tmpFirstNumber *10 + tmpSecondNubmer);
```
Comment (smay1613) [inline Lesson1/Denis Sulima/task2/main.cpp]: It is a good practice to extract such expressions into the named variable.

File: Lesson1/Denis Sulima/task3/main.cpp [hunk 5]
```diff
@@ -0,0 +1,24 @@
+#include <iostream>
+#include <forward_list>
+
+int main()
+{
+    std::forward_list<int>myForvardList{10,20,30,40,50};
+    std::forward_list<int>reverseForvardList;
+    std::cout <<"myForvardList:\t\t";
+    for(const auto &value: myForvardList){
```
Comment (smay1613) [inline Lesson1/Denis Sulima/task3/main.cpp]: Yep, that's right. Try to find alternative solutions, if you want.

File: Lesson1/Denis Sulima/task3/main.cpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        reverseForvardList.push_front(value);
+        std::cout<< value <<" ";
+        }
+    std::cout << std::endl <<"reverseForvardList:\t";
+    for(const auto &value: reverseForvardList){
+        std::cout<< value << " ";
+    }
+    myForvardList = reverseForvardList;
```
Comment (smay1613) [inline Lesson1/Denis Sulima/task3/main.cpp]: Alternatively, you can use assign(iterator, iterator) version.

**Comment by smay1613** [comment]
At first look, it is a nice work (I'll review it again after the changes in project structure). 
Please, can you extend the .gitignore file for excluding the *.txt.user files and merge all tasks into one project? 
The next task is to write unit tests for your functions.
Create the new sub-project (inside the merged one), called 'Unit tests'. In CMake, set up it with the Google Tests (you can use this article - https://www.testcookbook.com/book/cpp/setting-up-cmake-google-test.html).
Create three separated unit test classes. Then, create 5 tests for each task. Think about the corner cases (such as empty container, incorrect size passed, etc). 
Thanks.

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: Lesson1/Denis Sulima/task1/CMakeLists.txt.user [hunk 7]
```diff
@@ -0,0 +1,382 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (smay1613) [inline Lesson1/Denis Sulima/task1/CMakeLists.txt.user]: Delete this file from the repo manually

File: Lesson1/Denis Sulima/task1/main.cpp [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+int main()
+{
+    {
+//--Task1----------------------------------------------------------------------------------------------------------------------------
+    std::cout <<"Task 1-------------------------"<<std::endl;
```
Comment (smay1613) [inline Lesson1/Denis Sulima/task1/main.cpp]: Please write tests and export the task code to the corresponding functions without any input/output, just computing and returning the result.

File: Lesson1/Denis Sulima/task1/main.cpp [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+    std::cout << std::endl;
+
+    for (const auto &value : vectorFirstPlusSecond) {
+        std::cout << value << " ";
+    }
+    }
+//--Task2----------------------------------------------------------------------------------------------------------------------------
+  {
+    std::cout << std::endl<<std::endl<<"Task 2----------------------"<<std::endl;
+    int X;
```
Comment (smay1613) [inline Lesson1/Denis Sulima/task1/main.cpp]: Also, don't leave variables uninitialized.

File: Lesson1/Sulima Denis Homework 1/include/Utils.h [hunk 10]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+#include <string>
+#include <chrono>
```
Comment (smay1613) [inline Lesson1/Sulima Denis Homework 1/include/Utils.h:3]: These headers are not used here.

File: Lesson1/Sulima Denis Homework 1/include/Utils.h [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+#include <forward_list>
+#include <vector>
+
+
+
+
+namespace avprocessing
+{
+namespace utilities
```
Comment (smay1613) [inline Lesson1/Sulima Denis Homework 1/include/Utils.h:12]: You forgot to remove it :)

File: Lesson1/Sulima Denis Homework 1/src/Utils.cpp [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    }
+    std::vector<int> combineVectors(std::vector<int> firstVec, std::vector<int> secondVec){
+        std::vector<int>vectorFirst{firstVec};
+        std::vector<int>vectorSecond{secondVec};
+        std::vector<int>vectorFirstPlusSecond {combineVectors(vectorFirst.begin(), vectorFirst.end(),
+                                                              vectorSecond.begin(),vectorSecond.end())};
+        return vectorFirstPlusSecond;
+    }
+    //Task2 ---------------------------------------------------------------------------
+    size_t magicNumber(const size_t  m_number, const size_t  m_degree){
```
Comment (smay1613) [inline Lesson1/Sulima Denis Homework 1/src/Utils.cpp:26]: _Why "m"? In Hungarian notation, "m" prefix is used for class members._

File: Lesson1/Sulima Denis Homework 1/src/Utils.cpp [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+    std::vector<int> combineVectors(const std::vector<int>::iterator vectorFirstBegin, const std::vector<int>::iterator vectorFirstEnd,
+                                    const std::vector<int>::iterator vectorSecondBegin, const std::vector<int>::iterator vectorSecondEnd){
+        std::vector<int>tmp {vectorFirstBegin,vectorFirstEnd};
+        tmp.insert(tmp.end(), vectorSecondBegin, vectorSecondEnd);
+        return tmp;
+    }
+    std::vector<int> combineVectors(std::vector<int> firstVec, std::vector<int> secondVec){
+        std::vector<int>vectorFirst{firstVec};
+        std::vector<int>vectorSecond{secondVec};
+        std::vector<int>vectorFirstPlusSecond {combineVectors(vectorFirst.begin(), vectorFirst.end(),
```
Comment (smay1613) [inline Lesson1/Sulima Denis Homework 1/src/Utils.cpp:21]: _Please use spaces!_

File: Lesson1/Sulima Denis Homework 1/src/Utils.cpp [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+        return result;
+    }
+
+    //Task3 ---------------------------------------------------------------------------
+    void reverse(std::forward_list<int> & m_ForvardList){
+        std::forward_list<int>reverseForvardListData;
+        for(const auto &value: m_ForvardList){
+            reverseForvardListData.push_front(value);
+        }
+        const std::forward_list<int> returnlist{reverseForvardListData};
```
Comment (smay1613) [inline Lesson1/Sulima Denis Homework 1/src/Utils.cpp:69]: **Why did you copied it?**

File: Lesson1/Sulima Denis Homework 1/tests/unit_tests/headers/TestMetadataUtils.h [hunk 15]
```diff
@@ -0,0 +1,14 @@
+#pragma once
+#include "gtest/gtest.h"
+#include <forward_list>
+
+class ReverceForwardList : public ::testing::TestWithParam<std::forward_list<int>>
```
Comment (smay1613) [inline Lesson1/Sulima Denis Homework 1/tests/unit_tests/headers/TestMetadataUtils.h:5]: _Typo._

File: Lesson1/Sulima Denis Homework 1/tests/unit_tests/headers/TestMetadataUtils.h [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+{
+};
+class combineVectors : public ::testing::TestWithParam<std::pair<std::vector<int>, std::vector<int>>>
```
Comment (smay1613) [inline Lesson1/Sulima Denis Homework 1/tests/unit_tests/headers/TestMetadataUtils.h:8]: _PascalCase._

File: Lesson1/Sulima Denis Homework 1/tests/unit_tests/src/TestMetadataUtils.cpp [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+TEST_P(combineVectors, combineVectors)
+{
+std::vector<int> firstVec = std::get<0>(GetParam());
+std::vector<int> secondVec = std::get<1>(GetParam());
+const std::vector<int> result {avprocessing::utilities::combineVectors(firstVec.begin(), firstVec.end(),
+                                                                       secondVec.begin(), secondVec.end())};
+EXPECT_EQ(result, (std::vector<int>{1, 2, 3, 4, 5, 6}));
+}
+
+const size_t checkResultMagicNumber[]{19316911, 1010101010, 525156535, 24, 0, 3719, 7493321171983, 0, 1};
```
Comment (smay1613) [inline Lesson1/Sulima Denis Homework 1/tests/unit_tests/src/TestMetadataUtils.cpp:23]: **Don't use c-arrays!**
If you really shoud use array, use std::array instead.

File: Lesson1/Sulima Denis Homework 1/tests/unit_tests/src/TestMetadataUtils.cpp [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+                                          std::make_pair(size_t {0},  size_t{0}),
+                                          std::make_pair(size_t {37}, size_t{2}),
+                                          std::make_pair(size_t {7},  size_t{7}),
+                                          std::make_pair(size_t {10}, size_t{0}),
+                                          std::make_pair(size_t {1},  size_t{1})
+                                          ));
+
+TEST_P(MagicNumber, MagicNumber)
+{
+    const size_t number = std::get<0>(GetParam());
```
Comment (smay1613) [inline Lesson1/Sulima Denis Homework 1/tests/unit_tests/src/TestMetadataUtils.cpp:41]: Use .first and .second properties in pair.
Also, you can use std::tie for pair unpacking.

File: Lesson1/Sulima Denis Homework 1/tests/unit_tests/src/TestMetadataUtils.cpp [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+
+std::vector<std::forward_list<int>>checkResultForwardListReverce{{5, 4, 3, 2, 1},
+                                                                 {100, 200, 300, 400, 500},
+                                                                 {},
+                                                                 {1},
+                                                                 {50, 50}
+                                                                };
+size_t numbersForwardListTest {0};
```
Comment (smay1613) [inline Lesson1/Sulima Denis Homework 1/tests/unit_tests/src/TestMetadataUtils.cpp:55]: You can make it member of testing class.
Also, make it static. And use fixture tests for such cases, when you're counting something.

File: Lesson1/Sulima Denis Homework 1/tests/unit_tests/src/TestMetadataUtils.cpp [hunk 20]
```diff
@@ -0,0 +1,80 @@
+#include "TestMetadataUtils.h"
```
Comment (smay1613) [inline Lesson1/Sulima Denis Homework 1/tests/unit_tests/src/TestMetadataUtils.cpp:1]: Rename it.
